### PR TITLE
Extend `ActiveRecord::Core#strict_loading!` to accept a block

### PR DESCRIPTION
**Pull request by seanpdoyle**
### Motivation / Background

Once a record enables or disables strict loading, that change is permanent. By extending the `#strict_loading!` method to support a block, that change can be temporarily applied **only within the block**, then reverted.

### Detail

For parity, this commit also introduces a class-level `ActiveRecord::Core.strict_loading!` that configure `strict_loading_by_default` with the same support for a self-contained blocks:

Instance method:

```ruby
user = User.first
user.strict_loading! do
  user.comments # => raises ActiveRecord::StrictLoadingViolationError
end
user.comments   # => #<ActiveRecord::Associations::CollectionProxy>
```

Class method:

```ruby
User.strict_loading!(true) do
  user = User.first
  user.comments # => raises ActiveRecord::StrictLoadingViolationError
end
user = User.first
user.comments   # => #<ActiveRecord::Associations::CollectionProxy>
```

### Additional information

The newly introduced [Object#with](https://edgeapi.rubyonrails.org/classes/Object.html#method-i-with) would be perfect candidate for the self-contained configurations. Unfortunately:

1. Active Record classes define [ActiveRecord::QueryMethods#with](https://edgeapi.rubyonrails.org/classes/ActiveRecord/QueryMethods.html#method-i-with) for querying
2. Active Record instances store strict loading internally as instance variables

Without the `Object#with` abstraction, the implementation's setup and teardown is messy.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activerecord/lib/active_record/core.rb [hunk 1]
```diff
@@ -222,6 +222,56 @@ def self.connection_class_for_self # :nodoc:
...
+      #   user.comments
+      #   => #<ActiveRecord::Associations::CollectionProxy>
+      #
+      #   User.strict_loading!(true) # => true
+      #   user = User.first
+      #   user.comments
+      #      => ActiveRecord::StrictLoadingViolationError
+      def self.strict_loading!(value = true, &block)
+        original_strict_loading_by_default = strict_loading_by_default
+        self.strict_loading_by_default = value
```
Comment (p8) [inline activerecord/lib/active_record/core.rb:262]: As this is a class method I don't think this is threadsafe.

Comment (seanpdoyle) [reply activerecord/lib/active_record/core.rb:262]: That's a great point, thank you!

Is there a thread-safe mechanism to make that method viable, or do those issues make it a non-starter?

Comment (p8) [reply activerecord/lib/active_record/core.rb:262]: Maybe use Threadlocal, similar to: https://github.com/rails/rails/blob/12fc98fdbb1c5038c43aa8873d961cf5046505ed/activerecord/lib/active_record/suppressor.rb#L33-L37
See discussion here: https://github.com/rails/rails/issues/18847#issuecomment-73435451

Comment (p8) [reply activerecord/lib/active_record/core.rb:262]: I see searching for `IsolatedExecutionState` returns many more examples.

**Comment by eileencodes** [comment]
Hey @seanpdoyle thanks for this but I don't know if we want this change. Strict loading is designed to be more global, set it once and you're done. This would encourage applications to sprinkle strict loading around where they want when convenient but I feel like that defeats the purpose of the feature. Can you describe your use case for this feature?

**Comment by seanpdoyle** [comment]
> I feel like that defeats the purpose of the feature. Can you describe your use case for this feature?

@eileencodes appreciate this feedback. I think your assessment is 100% spot on. 

In my experience, there have been times where the self-contained nature of the setup-exercise-verification structure of tests cause strict loading violations.

For those tests,we define our own `with_strict_loading_disabled` test helper that accepts a block. We usually wrap the Verification portion of the test in that block so that we can retrieve records without much consideration for violations. Since the tests themselves are focused and start with a clean slate database, N+1 queries are not a big concern.

This proposal aimed to codify that behavior as part of the framework itself.

Now that you mention it, I don't recall ever using or wanting that behavior on the implementation side.

**Comment by seanpdoyle** [comment]
@eileencodes more recently, our project has bumped up against strict loading violations in our `ActiveJob::Base` descendants, mostly any job that accepts a Global ID-locatable resource as an argument to `#perform`. Consider a hypothetical scenario to demonstrate the friction.

An `Article` record (that's configured to enforce strict loading) and a `PublishArticleJob` that publishes the article and emails the authors to notify them:

```ruby
class PublishArticleJob < ApplicationJob
  def perform(article)
    # do some publishing logic

    # calling `article.authors` raises a violation
    article.authors.each { |author| notify(author) }
  end
end
```

Since `article` is loaded by a call to [GlobalID::Locator.locate](https://github.com/rails/rails/blob/v7.1.0.beta1/activejob/lib/active_job/arguments.rb#L136-L138), we have no access to the Active Record relation used to execute the query, so we cannot provide eager loading hints (by chaining `includes`, for example).

To work around this, we've called `article.strict_loading!(false)` atop all of our `#perform` definitions. It's easy to forget.

While the idea solution would be to provide Active Job with hooks to control how records are resolved during argument deserialization, a work-around that this PR proposes is to disable strict loading in a `around_perform` block.

Either something global like:

```ruby
around_perform do |job, block|
  ApplicationRecord.strict_loading!(false, &block)
end
```

or something specific like:

```ruby
around_perform do |job, block|
  Article.strict_loading!(false, &block)
end
```

**Comment by seanpdoyle** [comment]
I'm closing this in favor of exploring a resolution that's built-into Active Job: https://github.com/rails/rails/pull/50341.

**Pull request closed without merging**